### ISSUE TITLE
Only warn on typed wildcard usage in top-level type signatures

### ISSUE THREAD

**Issue by garyb** [purescript/purescript#1228]

**Comment by paf31** [comment]
This seems a bit too heavy-handed to me. I think of wildcards as a tool, and I might want to use a wildcard to infer a type inside a `let` binding, even though I agree that usually those tend to be the places where you don't care as much.

I think we might need a way to disambiguate between "I genuinely don't care what this type is" and "please infer this for me". One simple way might be to allow named wildcards:

```purescript
foo = do
  x <- get :: _state Int
```

where the expectation is that the warning should name the wildcard to help understand the inferred types:

```text
The type wildcard _state has the inferred type

    ExceptT e1 m1
```

where the usual `_` wildcard would take on the meaning "I don't care about the type, just infer it".

**Comment by garyb** [comment]
I was talking to Jon and/or Becky about this the other week actually, and came to the same conclusion.

I feel like the named wildcards should be the non-warned kind though: if you're giving them a name it's to give them some meaning, which means you can provide an idea of what it is your talking about. `_state` being a great example there, as that usage can explain what you're doing: "I'm inferring a state monad here, don't worry about it`, whereas `x <- get :: _ Int` gives you little clue to what `_` is without considering the wider context.

**Comment by garyb** [comment]
Didn't you suggest allowing named wildcards to do things like `_x -> _x` at one point too? That's something that `_` could never ever help you with.

**Comment by paf31** [comment]
Yes, that's the other issue with this - there is an orthogonal set of concerns - whether we name a wildcard to unify it with other wildcards.

All considered, I'm not sure naming is the right way to solve this issue, but I think it will be good to be explicit about which wildcards we want to print out.

**Comment by paf31** [comment]
We could deal with this and typed holes at the same time perhaps: use a typed hole when you want the type, and a wildcard when you don't:

```purescript
names = map getName? people
```

would give an error (not warning) along the lines of

```text
The hole getName? has the inferred type

  Person -> String
```

**Comment by garyb** [comment]
That would be great for most cases but I think there's still value in being able to get information out of a wildcard in a type too (I've used `Natural Something _` kind of thing a few times while working on/in Halogen). Perhaps we could use something similar there too, `?what -> Unit`.

I think we should still always warn on top-level wildcards still too, as that's like leaving the type off from the perspective of reading the code.

Maybe this is a crazy idea, but perhaps we don't need special syntax for holes: we could treat any unbound identifier as a hole - so you still get an error about it being undefined, but we could state the inferred type in that error.

**Comment by paf31** [comment]
Moving to 0.9, until we figure out syntax.

**Comment by paf31** [comment]
Moving this back to Ideas until we have a concrete proposal for syntax.

**Comment by garyb** [comment]
I'd suggest we adopt something similar to holes: `?foo` in a type always werrors, `_` warns in top level but not in where / let / expression annotations.

**Comment by garyb** [comment]
Resolved by #3500
